### PR TITLE
Improvement of InMemoryRoleProvider

### PR DESCRIPTION
**Pull request by Wilt**
This PR follows on the issue [here](https://github.com/ZF-Commons/zfc-rbac/issues/322)

A `$roles` array storage is added to the InMemoryRoleProvider in which resolved roles are stored. It means role instances with the same name are reused (and not recreated). This will save memory and allows 1 to 1 role object comparison since the `InMemoryRoleProvider` will always return the same object instance for a certain `$roleName`.

A test for this enhancement is provided.

### CODE REVIEW

**Comment by Wilt** [comment]
Failing tests are not because of my changes...

**Comment by Wilt** [comment]
Failing tests are not because of my changes...

File: src/Role/InMemoryRoleProvider.php [hunk 1]
```diff
@@ -62,19 +72,35 @@ public function getRoles(array $roleNames)
...
+     */
+    protected function getRole($roleName)
+    {
+        if (isset($this->roles[$roleName])) {
+            return $this->roles[$roleName];
+        }
+
+        // If no config, we create a simple role with no permission
+        if (!isset($this->rolesConfig[$roleName])) {
+            $role = new Role($roleName);
```
Comment (danizord) [inline src/Role/InMemoryRoleProvider.php]: Return early and avoid the else keyword

File: test/Role/InMemoryRoleProviderTest.php [hunk 2]
```diff
@@ -64,5 +64,8 @@ public function testInMemoryProvider()
         $this->assertEquals('guest', $guestRole->getName());
         $this->assertFalse($guestRole->hasPermission('write'));
         $this->assertFalse($guestRole->hasPermission('delete'));
+
+        $this->assertTrue($adminRole->getChildren()['member'] === $memberRole);
+        $this->assertTrue($memberRole->getChildren()['guest'] === $guestRole);
```
Comment (danizord) [inline test/Role/InMemoryRoleProviderTest.php]: Can use assertSame instead

**Comment by danizord** [comment]
Looks good for me, :+1:

@bakura10?

**Comment by Wilt** [comment]
@danizord Updated code following your comments

**Comment by bakura10** [comment]
Travis tests are not passing anymore. Actually you have made this PR against the develop branch which is a new major version that relies on a new version of Rbac.

I think that what you want is doing the PR against master.

**Comment by Wilt** [comment]
@bakura10 Sure, I never seem to do it right. Others want `develop` as target branch :S
I will change it...

**Comment by Wilt** [comment]
Now here: https://github.com/ZF-Commons/zfc-rbac/pull/325

**Pull request closed without merging**